### PR TITLE
Enable multi-platform support for operator image builds

### PR DESCRIPTION
**Pull request by ashokpariya0**
These changes enable building and pushing CNAO operator image for multiple platforms (amd64, s390x, arm64) from a single Dockerfile. Enhanced multi-platform support in the build process by adding a PLATFORMS argument in the Makefile for amd64, s390x, and arm64 architectures. Updated the docker-build-operator target to support builds with docker buildx and podman, and added new targets for creating Docker Buildx builders and pushing multi-platform images.

**What this PR does / why we need it**:
The existing CNAO operator ([link](https://quay.io/repository/kubevirt/cluster-network-addons-operator?tab=tags&tag=latest)) currently supports only the amd64 architecture. These changes extend multi-platform support to include arm64 and s390x architectures, enabling broader compatibility for the CNAO operator image.

**Special notes for your reviewer**:
Please note that Docker and Podman handle multi-architecture image builds differently. As a result, the changes have been made accordingly. Additionally, Go cross-compilation support has been added to build binaries for the target architecture.
Added a PLATFORMS argument in the Makefile to support building images for multiple architectures (amd64, s390x, arm64).
Updated the Dockerfile to dynamically handle multi-architecture builds.
Added Go cross-compilation support in Dockerfile to build binaries for the target architecture.
Enhanced the docker-build-operator target to enable multi-platform builds using both docker buildx and podman.
Introduced a new build-multiarch-operator-image target to build and push multi-platform images using podman.
Added a create-builder target to generate Docker Buildx builders for multi-platform image support.

**Release note**:
```release-note
None
```

### CODE REVIEW

File: .dockerignore
Comment (oshoval) [inline .dockerignore:15]: leftovers?

File: Makefile
Comment (oshoval) [inline Makefile:16]: lets have just amd64 by default, so it wont take time on each regular use ?
the rest can be in an example comment

moreover if we build just one image, we dont need buildx just docker build, much easier to maintain, usually this will be the case, unless on releases and such once time come

File: Makefile
Comment (oshoval) [inline Makefile:22]: can we just inline those please?

File: Makefile
Comment (oshoval) [inline Makefile:133]: please consider moving it under create-builder target
unless you think it doesnt fit there?

File: Makefile
Comment (oshoval) [inline Makefile:258]: please drop the comments (all over unless it is not trivial)

File: build/operator/Dockerfile
Comment (oshoval) [inline build/operator/Dockerfile:3]: might worth to auto take it from go.mod
there are few network repos that doing something like that

File: build/operator/Dockerfile
Comment (oshoval) [inline build/operator/Dockerfile:12]: maybe using docker ADD can simplify some of this ?

File: build/operator/Dockerfile
Comment (oshoval) [inline build/operator/Dockerfile:48]: please drop comments all over unless non trivial

File: hack/build-multiarch-operator.sh
Comment (oshoval) [inline hack/build-multiarch-operator.sh:3]: please drop comments unless non trivial

**Review by oshoval — COMMENTED** [review]
Nice thanks
First pass

File: hack/build-multiarch-operator.sh
Comment (oshoval) [inline hack/build-multiarch-operator.sh:1]: might worth to consider doing as close to 
https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh please

File: Makefile
Comment (ashokpariya0) [reply Makefile:16]: @oshoval Thanks for the feedback! I agree that simplifying the default platform to just amd64 will streamline the build process and reduce unnecessary overhead. I'll update the configuration accordingly and move the other platforms into an example comment as you suggested.

As for the use of docker build vs. buildx, that approach makes sense for most cases. I'll switch to docker build for simplicity and reserve buildx for when it's needed.

During releases, we'll set PLATFORMS ?= linux/amd64,linux/s390x,linux/arm64, use buildx, and build and push a multi-platform image for the operator. For all other cases, we'll default to using amd64 as the platform. Does that sound good?

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:3]: Good suggestion! I'll look into implementing that.

File: .dockerignore
Comment (ashokpariya0) [reply .dockerignore:15]: No, this change is necessary because we need to build the two binaries listed [here](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L107-L111) inside the Dockerfile in order to support Go cross-compilation for building the target binaries.

Comment (oshoval) [reply .dockerignore:15]: lets remove the lines then completely ?

File: Makefile
Comment (oshoval) [reply Makefile:16]: sgtm thx

File: .dockerignore
Comment (ashokpariya0) [reply .dockerignore]: Yes, Removed in the latest commit.

File: Makefile
Comment (ashokpariya0) [reply Makefile]: Updated in the latest commit.

File: Makefile
Comment (ashokpariya0) [reply Makefile:274]: Done.

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile]: Done

File: hack/build-multiarch-operator.sh
Comment (ashokpariya0) [reply hack/build-multiarch-operator.sh]: Done.

File: hack/build-multiarch-operator.sh
Comment (ashokpariya0) [reply hack/build-multiarch-operator.sh:1]: Updated in latest commit.
FYI- It would be preferable to pass the builder name at line 30 in the script [init-buildx.sh](https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh#L30).

**Comment by ashokpariya0** [comment]
@oshoval  The job [here](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/11794285408/job/32851563920?pr=1923) failed with the following error:

make[2]: Leaving directory '/tmp/cnao/cluster-network-addons-operator'
**podman manifest push --tls-verify=false localhost:5000/cluster-network-addons-operator:latest
Error: accepts 2 arg(s), received 1**
make[1]: *** [Makefile:143: docker-push-operator] Error 125
make[1]: Leaving directory '/tmp/cnao/cluster-network-addons-operator'
make: *** [Makefile:168: cluster-operator-push] Error 2
Issue Description:

The error indicates that the podman manifest push command is receiving only one argument (localhost:5000/cluster-network-addons-operator:latest), but it expects two arguments (the manifest and the destination).

The command works as expected in my local environment, where I'm using podman version 4.9.4-rhel.
It seems that the issue arises in the CI pipeline, which may be using a different version of Podman, leading to different behavior. Any thoughts?

**Comment by oshoval** [comment]
@phoracek @RamLavi @0xFelix 
please consider also reviewing if you can, to see i dont miss anything

File: hack/build-multiarch-operator.sh
Comment (oshoval) [reply hack/build-multiarch-operator.sh:1]: can we make it even closer to that script ?
will be easier to maintain, it still seems too close to the original version if i am not wrong

File: Makefile
Comment (oshoval) [reply Makefile:274]: please remove more comments, there are still some trivial ones
we usually add as few as possible
only for non trivial stuff

File: Makefile
Comment (oshoval) [inline Makefile:147]: please try to think about a makefile target structure that can be simpler wrt to multi platform and one platform
the logic is split here and on docker-build-operator, it will be better to have it concentrated if possible (i understand it might be not possible because there are stages)

**Review by oshoval — COMMENTED** [review]
Thanks,
there are some comments that were not addressed / replied to
please check those as well

File: Makefile
Comment (ashokpariya0) [reply Makefile:147]: For docker buildx, we need to combine both build and push in a single command because the images are not stored locally by default, unlike with docker build. As a result, we don't need a separate docker push command when working with multi-platform builds. To simplify the logic and maintain consistency, we could consider adding the --push option to docker build as well. What do you think @oshoval

Comment (oshoval) [reply Makefile:147]: might be interesting to try if you think it will simplify it please

File: hack/build-multiarch-operator.sh
Comment (ashokpariya0) [reply hack/build-multiarch-operator.sh:1]: Yes, I did that.
Just to let you know, I noticed an issue in [this line](https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh#L30) of init-buildx.sh — we are not performing the inspection based on the builder name.

File: Makefile
Comment (ashokpariya0) [reply Makefile]: Done.

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile]: I made the changes according to the script [install-go.sh](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/hack/install-go.sh). With the current setup, we are compiling and building the manager binary using the Go version specified above.

Please note that the script [go-version.sh](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/hack/go-version.sh) returns the Go version in major.minor format. Additionally, the command curl -L -s "https://go.dev/dl/?mode=json" does not list Go 1.21, as it's archived. Before Go 1.22, we would need to download it directly from the archive. However, the install-go.sh script currently downloads the latest Go version and proceeds to compile with it.

I’ve made the same update in the Dockerfile to ensure consistency.

File: hack/build-multiarch-operator.sh
Comment (oshoval) [reply hack/build-multiarch-operator.sh:1]: thanks, if you think it worth it you can open an issue on nmstate repo
i must work on some other stuff at this very moment, so will take me bit time to review

**Review by 0xFelix — COMMENTED** [review]
Perhaps an unpopular opinion, but why do we need to maintain two builds for Podman and Docker? In https://github.com/kubevirt/ssp-operator we went with Podman only.

See https://github.com/kubevirt/ssp-operator/pull/1091

**Comment by oshoval** [comment]
> Perhaps an unpopular opinion, but why do we need to maintain two builds for Podman and Docker? In https://github.com/kubevirt/ssp-operator we went with Podman only.
> 
> See [kubevirt/ssp-operator#1091](https://github.com/kubevirt/ssp-operator/pull/1091)

personally i am against dropping docker support
it is convenient it exists, and also on kubevirt we still do that for some jobs
moreoever, cnao depends on lot of other component, so it better to be robust in case something happens and is required

While it is a nice question, not sure if the PR is the right place for it (i.e it shouldn't block it)

File: Makefile
Comment (0xFelix) [inline Makefile:27]: Are these vars really required?

File: Makefile
Comment (0xFelix) [inline Makefile:142]: This is called docker-push-operator but uses podman?

File: Makefile
Comment (0xFelix) [inline Makefile:256]: Can you reuse/deduplicate the args?

File: hack/build-multiarch-operator.sh
Comment (0xFelix) [inline hack/build-multiarch-operator.sh:9]: Hardcoded buildx version?

File: hack/build-multiarch-operator.sh
Comment (0xFelix) [inline hack/build-multiarch-operator.sh:1]: Did you run shellcheck against this file?

**Review by 0xFelix — COMMENTED** [review]
@oshoval Not to block the PR, but IMO it adds a lot of complexity which I'm not sure is required. (especially looking at `hack/build-multiarch-operator.sh`)

**Comment by oshoval** [comment]
> @oshoval Not to block the PR, but IMO it adds a lot of complexity which I'm not sure is required. (especially looking at `hack/build-multiarch-operator.sh`)

we can indeed request to have a separate script for it instead having it internal as it is now, wdyt?

File: Makefile
Comment (ashokpariya0) [reply Makefile]: It was needed, but removed it latest commit and moved to shell script.

File: Makefile
Comment (ashokpariya0) [reply Makefile]: Updated in latest commit.

File: hack/build-multiarch-operator.sh
Comment (ashokpariya0) [reply hack/build-multiarch-operator.sh]: good catch, corrected it.

File: hack/build-multiarch-operator.sh
Comment (ashokpariya0) [reply hack/build-multiarch-operator.sh:1]: Yes.

**Comment by ashokpariya0** [comment]
@oshoval @0xFelix The review comments have been addressed. Please let me know if there's anything else I need to do.

File: build/operator/Dockerfile
Comment (0xFelix) [inline build/operator/Dockerfile:16]: Is this arg used?

File: hack/build-operator-docker.sh
Comment (0xFelix) [inline hack/build-operator-docker.sh:21]: Empty lines at EOF

File: build/operator/Dockerfile
Comment (0xFelix) [inline build/operator/Dockerfile:15]: Where are these set?

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:16]: Yes

File: hack/build-operator-docker.sh
Comment (ashokpariya0) [reply hack/build-operator-docker.sh]: Removed.

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:15]: TARGETOS and TARGETARCH are set by docker/podman based on the platform you're building for based on (--platform ) flag.

**Comment by ashokpariya0** [comment]
@oshoval @0xFelix  I’ve addressed all the review comments.
Could you please take another look and proceed with the next steps for the PR?
Let me know if there’s anything else you’d like me to update.

File: Makefile
Comment (0xFelix) [reply Makefile:138]: @ashokpariya0 Can you comment on this?

File: build/operator/Dockerfile
Comment (0xFelix) [reply build/operator/Dockerfile:16]: I may be blind, but where?

File: Makefile
Comment (ashokpariya0) [reply Makefile:138]: Yes, We may need to modify the section at Makefile lines 113-126(https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L113-L126), but this can be addressed in a separate PR, as renaming might impact some build processes or jobs. The existing code already supports both Docker and Podman, as indicated in [Makefile line 14](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L14).

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:16]: we use at https://github.com/kubevirt/cluster-network-addons-operator/blob/106c2b661dff49532a0e2b5738f7c12be41a48f2/build/operator/Dockerfile#L18-L19

Comment (ashokpariya0) [reply build/operator/Dockerfile:16]: TARGETPLATFORM can be removed.

File: Makefile
Comment (0xFelix) [reply Makefile:138]: With your change this target will only push in case `OCI_BIN` is equal to `podman`, IIUC?

Comment (ashokpariya0) [reply Makefile:138]: Yes exactly, with docker buildx, push is combined with image build.

Comment (0xFelix) [reply Makefile:138]: Fine for me, @oshoval wdyt?

Comment (oshoval) [reply Makefile:138]: fine by me to address naming on follow-up
thx

**Review by 0xFelix — APPROVED** [review]
/lgtm

File: build/operator/Dockerfile
Comment (oshoval) [inline build/operator/Dockerfile:6]: i dont think we want unpinned go version?
stuff can break suddenly

unless we are fine with that or if this was the case already, lets see other maintainers idea

File: hack/build-operator-docker.sh
Comment (oshoval) [inline hack/build-operator-docker.sh:8]: please drop comments
from all files, unless non trivial

**Review by oshoval — COMMENTED** [review]
quick pass

**Comment by oshoval** [comment]
please see pull-cluster-network-addons-operator-unit-test

**Comment by ashokpariya0** [comment]
> please see pull-cluster-network-addons-operator-unit-test
>
> Sure.

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:6]: This was already the case; this PR follows the same approach.

File: Makefile
Comment (ashokpariya0) [reply Makefile:138]: @oshoval  I've already looked into this and found that `docker-*` is being called from various scripts in the automation, hack folder, etc. Additionally, we need to update the README and other related files. Including all of this in a single commit would add unnecessary complexity. It would be better to handle these changes separately.

Comment (oshoval) [reply Makefile:138]: ok thanks

**Comment by oshoval** [comment]
Thanks

@qinqon 
any thing you want to add please ?

File: Makefile
Comment (ashokpariya0) [reply Makefile:138]: @oshoval I will create an separate issue for this, resolve it, and provide a follow-up.

File: build/operator/Dockerfile
Comment (ashokpariya0) [reply build/operator/Dockerfile:6]: @oshoval For this as well, I will create a separate issue, resolve it, and provide a follow-up.

Comment (oshoval) [reply build/operator/Dockerfile:6]: thanks, if this was already the case, no rush to fix it, its ok

**Comment by ashokpariya0** [comment]
Could we please have a release once this PR is merged? There were commits made after the last release (v0.96.0) on October 8th. We would like to perform end-to-end testing on this release.
Also, as discussed earlier, please set PLATFORMS=linux/amd64,linux/s390x before pushing the upstream image, as by default it will only publish amd64-compatible images.

**Comment by oshoval** [comment]
yes thanks but at this very moment i cant work on that
once one of us have more time

**Comment by ashokpariya0** [comment]
> yes thanks but at this very moment i cant work on that once one of us have more time
> 
> you are welcome if you want to check project-infra and suggest a PR
>
> Okay, thanks! For now, merging the PR once it passes the tests would be helpful.

**Comment by oshoval** [comment]
@RamLavi 
if you want to also take a look, before we approve soon

**Comment by ashokpariya0** [comment]
@RamLavi Would like to add anything ??
@oshoval Could you please approve the changes if no further actions are required, so that it can trigger the other tests?

File: build/operator/Dockerfile
Comment (oshoval) [reply build/operator/Dockerfile:6]: Please consider in the follow-up to use ADD if it helps

**Comment by oshoval** [comment]
Very nice thanks
/approve

**Comment by ashokpariya0** [comment]
@oshoval Thanks for approving the PR! How can we trigger all the tests for it?

**Comment by oshoval** [comment]
/test all

thought it auto triggers once lgtm approve

**Comment by ashokpariya0** [comment]
Thanks @oshoval  all tests have passed. There's one pending test job (tide) that requires the LGTM label.

**Comment by oshoval** [comment]
Hi @0xFelix 
Can you please review / lgtm once you are good with it ?
Thanks

**Review by 0xFelix — APPROVED** [review]
/lgtm

Thanks!

**Comment by oshoval** [comment]
Hi
Please check this error before we advance on any PR

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/1923/pull-cluster-network-addons-operator-unit-test/1860989134017925120#1:build-log.txt%3A10836

also i dont understand how come some lanes seems they didn't run
I will create a vanilla main PR to see CNAO state

**Comment by oshoval** [comment]
I think it will be the best to revert this PR and recreate it (due to the CI problems above)
I will also need time to update my own machine to see it works locally

File: hack/build-operator-podman.sh
Comment (oshoval) [inline hack/build-operator-podman.sh:18]: please drop the no-cache on a follow-up

File: hack/build-operator-podman.sh
Comment (oshoval) [inline hack/build-operator-podman.sh:12]: please consider adding redir of stderr to /dev/null so it will clean up the logs from marking error on CI

File: hack/build-operator-podman.sh
Comment (ashokpariya0) [reply hack/build-operator-podman.sh:18]: Done.

File: hack/build-operator-podman.sh
Comment (ashokpariya0) [reply hack/build-operator-podman.sh:12]: Added in PR: https://github.com/kubevirt/cluster-network-addons-operator/pull/1928

File: hack/build-operator-podman.sh
Comment (oshoval) [reply hack/build-operator-podman.sh:18]: Thanks

**Pull request merged**
